### ISSUE TITLE
Electronic Speech System

### ISSUE THREAD

**Issue by hipco** [dirkwhoffmann/virtualc64#785]
After V4.5 Ghostbusters and Impossible Mission have a distorted Electronic Speech System output. I've tried many different settings but none of them worked for me. Is there something I'm missing in the machine configuration settings or is it a bug?

**Comment by mithrendal** [comment]
Did you select the 6581 SID? The newer 8580 does not so well at speach synthesis. Commodore removed a bug in the newer SID which was exploited for speach synthesis.

**Comment by dirkwhoffmann** [comment]
Confirmed. Something is broken.

**Comment by hipco** [comment]
Yup, bug. SID 6581 doesn't make any difference.

**Comment by dirkwhoffmann** [comment]
Seems to work in v4.6. First broken release is v4.7b1

**Comment by hipco** [comment]
Correct. V4.6 as stated with my first post works.

**Comment by dirkwhoffmann** [comment]
Did a binary search on the repo: Culprit is commit 55aadf4.

**Comment by dirkwhoffmann** [comment]
Culprit is the CIA sleep logic. Here is the relevant part with an additional if-statement added for debugging: 

```C++
    // Sleep if threshold is reached
    if (tiredness > 8 && !CIA_ON_STEROIDS) {
        sleep();
        scheduleWakeUp();

        // EXPERIMENTAL TO HUNT DOWN #785
        if (wakeUpCycle <= sleepCycle) {

            printf("Alert: %lld %lld\n", wakeUpCycle, sleepCycle);
            wakeUp();
        }

    } else {
        scheduleNextExecution();
    }
```

During speech synthesis, the following output is produced: 

```
Alert: 60326377 60326377
Alert: 60331070 60331070
Alert: 60338006 60338006
...
```
The CIA puts itself to sleep with a wake-up cycle that matches the current clock. In the old code, events were processed after each half-cycle which woke the CIA in the same cycle it fell asleep. In the optimized code (with events only processed once per cycle), the wakeup comes one CPU cycle too late, resulting in a stuttered sound.

Clean fix: Alter the sleep logic such the CIA is only put to rest if it sleeps for at least 'n' cycles (with a suitable 'n').

**Comment by dirkwhoffmann** [comment]
Here is a small speed-improvement idea. Current code: 
```C++
void
CIA::serviceEvent(EventID id)
{
    switch(id) {

        case CIA_EXECUTE:

            executeOneCycle();
            break;

        case CIA_WAKEUP:

            wakeUp();
            break;

        default:
            fatalError;
    }
}
```
Idea: Since `CIA_WAKEUP` is a rare event, emulation can be sped up by adding a separate `CIA_WAKEUP_SLOT` in the secondary event table. By doing so, the `CIA_EVENT` slot would only be responsible for CIA execution. The dispatch function above could be eliminated, and `executeOneCycle()` could be called directly whenever a CIA event fires.

**Comment by hipco** [comment]
I'm impressed. Suitable 'n' would be > 1 in the speech synthesis scenario?

**Comment by hipco** [comment]
Any news about V4.8?

**Comment by dirkwhoffmann** [comment]
Fixed in v5.0b1
